export const fromPromise_ = (f, onError) => (_, sink) => {
    let promise;
    let aborted = false;
    sink(0, (signal) => {
        if (signal === 1) {
            if (promise)
                return;
            try {
                promise = f()
                    .then((a) => {
                    if (aborted)
                        return;
                    sink(1, a);
                    sink(2, undefined);
                })
                    .catch((err) => {
                    if (aborted)
                        return;
                    sink(2, onError(err));
                });
            }
            catch (err) {
                sink(2, onError(err));
            }
        }
        else if (signal === 2) {
            aborted = true;
        }
    });
};
export const fromPromise = (onError) => (f) => fromPromise_(f, onError);
